### PR TITLE
Impute missing values

### PR DESCRIPTION
**Pull request by gbrunin**
Some features can be NaN after featurization by matminer or by the user (through df_featurized). This should be handled by MODNet (by matminer as well, but that's another story). It already is handled at different places but, in the case of featurization through MODNet, [the infinite and NaN features are replaced by zeros.](https://github.com/ppdebreuck/modnet/blob/master/modnet/preprocessing.py#L772) This means that the various np.nan_to_num(x) calls later in the code are not useful in most cases, and the strategy that has been adopted that consists in setting NaNs to -1 after scaling x between -0.5 and 0.5 is never used. 

In this PR, I try to address these issues. After the featurization, the NaNs are not replaced by 0 anymore. **This breaks the tests as some features are slightly different.** The infinite values are replaced by NaNs. Then, the NaNs are handled when fitting the model using a [SimpleImputer](https://scikit-learn.org/stable/modules/generated/sklearn.impute.SimpleImputer.html#sklearn.impute.SimpleImputer) which can be chosen. It is then stored as an attribute to the model, and can be re-used when predicting new values. The scaler can also be chosen (StandardScaler or MinMaxScaler), and the user can also choose to first impute then scale, or first scale then impute. Both can be argued (do we want to keep the same distribution as the initial feature, or to change it by moving the NaNs outside the distribution). 

One thing to test is whether saving and loading models is working with the imputer and scaler (I just discussed this with PP, it needs to be checked).

### CODE REVIEW

File: modnet/hyper_opt/fit_genetic.py [hunk 1]
```diff
@@ -33,25 +34,35 @@ def __init__(
             weights (Dict[str, float]): Optional (for joint learning only). The relative loss weights to apply for each target.
         """
 
+        self.act = "elu"
+        self.loss = "mae"
+        self.n_neurons_first_layer = 32 * random.randint(1, 10)
```
Comment (ml-evs) [inline modnet/hyper_opt/fit_genetic.py:39]: I wonder if we should use a fixed/customisable seed here so that fit genetic can be made somewhat deterministic?

File: modnet/hyper_opt/fit_genetic.py [hunk 2]
```diff
@@ -477,7 +496,7 @@ def run(
             n_jobs (Optional[int], optional): Number of jobs to parallelize on. Defaults to None.
             early_stopping (Optional[int], optional): Number of successive generations without improvement before stopping. Defaults to 4.
             refit (Optional[int], optional): Wether to refit (>0) the best hyperparameters on the whole dataset or use the best Individual instead (=0).
```
Comment (ml-evs) [inline modnet/hyper_opt/fit_genetic.py:498]:
```suggestion
            refit (Optional[int], optional): Whether to refit (>0) the best hyperparameters on the whole dataset or use the best Individual instead (=0).
```

File: modnet/models/vanilla.py [hunk 3]
```diff
@@ -237,6 +245,17 @@ def fit(
...
+                If str, defines the strategy used in the scikit-learn SimpleImputer,
+                e.g., "mean" sets the NaNs to the mean of their feature column.
+                If a float is provided, and if xscale_before_impute is False, this
+                float is used to replace NaNs in the original dataset.
+                If a float is provided but xscale_before_impute is True, the float
+                is not used and standard values are used.
+                If you want to do something more sophisticated, make your own
+                modifications to MODData.df_featurized before fitting the model.
+            xscale_before_impute: whether to first scale the input and then impute values, or
+                                  first impute values and then scale the inputs.
```
Comment (ml-evs) [inline modnet/models/vanilla.py:258]:
```suggestion
            xscale_before_impute: whether to first scale the input and then impute values, or
                first impute values and then scale the inputs.
```

**Review by ml-evs — COMMENTED** [review]
Minor comments only in the code review, I haven't been able to test it out on any data yet but should have a chance to before we merge it

File: modnet/hyper_opt/fit_genetic.py [hunk 4]
```diff
@@ -33,25 +34,35 @@ def __init__(
             weights (Dict[str, float]): Optional (for joint learning only). The relative loss weights to apply for each target.
         """
 
+        self.act = "elu"
+        self.loss = "mae"
```
[the remark this answers is not in the document]
Comment (gbrunin) [reply modnet/hyper_opt/fit_genetic.py:38]: Would this make it deterministic though? Since the models will in any case lead to somewhat slightly different results each time they are fitted, they might differ from iteration to iteration?

File: modnet/hyper_opt/fit_genetic.py (hunk 1 above)
Comment (gbrunin) [reply modnet/hyper_opt/fit_genetic.py:39]: Would this make it deterministic though? Since the models will in any case lead to somewhat slightly different results each time they are fitted, they might differ from iteration to iteration?

**Review by ml-evs — APPROVED** [review]
LGTM!

semble bon pour moi

File: modnet/hyper_opt/fit_genetic.py [hunk 5]
```diff
@@ -646,7 +660,7 @@ def run(
 
         else:
             ensemble = []
-            for m in models[ranking[:10]]:
+            for m in models[ranking[:refit]]:
```
Comment (ppdebreuck) [inline modnet/hyper_opt/fit_genetic.py:663]: Why ? 10 is arbitrary sure... But now, refit=0 is throwing an error unfortunately...

Comment (gbrunin) [reply modnet/hyper_opt/fit_genetic.py:663]: Ah, oops. Shouldn't it be just the first model though?
